
# Profitable App Profiles for the App Store and Google Play Markets


The aim of this project is to find mobile app profiles that are profitable for the App Store and Google Play markets. 

A company that builds an Android and iOS mobile apps want to make their apps available on Google Play and the App Store that are free to download and install. The main source of revenue are in-app ads. This means that the revenue for any given app is mostly influenced by the number of users who use the app — the more users that see and engage with the ads,the better. My goal for this project is to analyze the data to help the company understand what type of apps are likely to attract more users.



## 1.Opening and Exploring the Data

By September 2018, there were approximately 4 million apps in the App Store and Google Play market.Collecting data for over 4 million apps requires a significant amount of time and money, so I will try to find a sample of data to analyze it. Luckily, these are two data sets that seem suitable for our purpose:

1- [App Store data set](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps): containing 7,000 iOS apps. You can download the data set directly from this [link](https://dq-content.s3.amazonaws.com/350/AppleStore.csv) 

2- [Google Play data set](https://www.kaggle.com/lava18/google-play-store-apps): containing 10,000 Android apps. You can download the data set directly from this [link](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv).


Let's start by opening the data sets:

In [8]:
from csv import reader

#AppStore data
opened_file = open('AppleStore.csv')
read_file = reader(opened_file)
appStore_data = list(read_file)
appStore_Header = appStore_data[0]
appStore_data = appStore_data[1:]

#GooglePlay data
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
googlePlay_data = list(read_file)
googlePlay_Header = googlePlay_data[0]
googlePlay_data = googlePlay_data[1: ]

To make it easier to explore the data sets, we need to write a function named **explore_data()** that we can use repeatedly to explore rows in a more readable way. The function takes the following parameters: 
 * The **dataset** - expected to be a list of lists.
 * Index of **start** and **end** row - expected to be integers. 
 * **rows_and_columns** to show the number of row and column - expected to be a Boolean and has False as a default argument.


In [9]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n')  

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))


In [10]:
print(appStore_Header)
print('\n')
explore_data(appStore_data, 0, 4, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


Number of rows: 7197
Number of columns: 16


- We have 7197 iOS apps in App Store data set.
- The columns that seem interesting are: 'track_name','price', 'rating_count_tot','prime_genre'.  

In [11]:
print(googlePlay_Header)
print('\n')
explore_data(googlePlay_data ,0 ,4, True ) 

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 10841
Number of columns: 13


- We have 10841 Android apps in Google Play data set.
- The columns that seem interesting are: 'App', 'Category', 'Rating', 'Reviews', 'Installs', 'Type', 'Price', 'Genres'.  

## 2.Data Cleaning 

After exploring the data,we need to start preparing and cleaning our data for analysis. 

Recall that the company only build free-apps and only for English-speaking audience This means that we'll need to remove: non-English apps & the apps that aren't free.

The cleaning process includes:

1- Detect inaccurate data, and correct or remove it.

2- Detect duplicate data, and remove the duplicates.

3- Remove non-English apps like 爱奇艺PPS -《欢乐颂2》电视剧热播.

4- Remove apps that aren't free.

### 2.1 Deleting Wrong Data

If we check the discussion forum on Google Play data set, we can see that one of the discussions [here](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) describes an error for row 10472 (without counting header row). Let's print this row and compare it against the header.

In [12]:
print(googlePlay_Header)
print('\n')
print(googlePlay_data[10472])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


The row 10472 corresponds to the app 'Life Made WI-Fi Touchscreen Photo Frame' is wrong because, the rating is 19 while the maximum value of app rating on Google Play is 5. So, we need to delete this data.


In [13]:
# don't run this more than once 
print(len(googlePlay_data))
del googlePlay_data[10472]
print(len(googlePlay_data))

10841
10840


### 2.2 Removing Duplicate Entries

If you explore the Google Play data set long enough, we'll find that some apps have more than one entry. So, we need to check what duplicate applications we have in the data set and how many of them.

In [14]:
duplicate_apps = []
unique_apps = []

for app in googlePlay_data:
    app_name = app[0]
    
    if app_name in unique_apps:
        duplicate_apps.append(app_name)
    else:
        unique_apps.append(app_name)


In [16]:
print("Number of unique apps" , len(unique_apps) )
print("\n")
print("Number of duplicate apps" ,  len(duplicate_apps) )

Number of unique apps 9659


Number of duplicate apps 1181


In total, there are 1,181 cases where an app occurs more than once.
Below, we print some of duplicate apps:

In [145]:
print("Example of duplicate apps: \n" , duplicate_apps[:10])

Example of duplicate apps: 
 ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


If we check how many entries we have for 'slack' app:

In [18]:
for app in googlePlay_data:
    if app[0] == 'Slack':
        print('\n')
        print(app)



['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']


['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']


['Slack', 'BUSINESS', '4.4', '51510', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']


If you examine the rows we printed for 'Slack' app, the main difference happens on index 3 of each row, which corresponds to the number of reviews. The different numbers show the data was collected at different times.

We can use this information to build a criterion for removing the duplicates.The higher the number of reviews, the more recent the data should be. Rather than removing duplicates randomly, we'll only keep the row with the highest number of reviews and remove the other entries for any given app.

We need to create a dictionary where: 
each key is a unique app name and the value is the highest number of reviews of that app.

In [19]:
reviews_max = {} # contains all unique apps names with their highest number of reviews.

for app in googlePlay_data:
    name = app[0]
    n_reviews = float(app[3]) 
    
    if (name in reviews_max) and (reviews_max[name] < n_reviews):
        reviews_max[name] = n_reviews
        
    elif name not in reviews_max :
         reviews_max[name] = n_reviews


Previously, we found that there are 1,181 duplicate apps, so the length of our dictionary (of unique apps) should be equal to the difference between the length of our data set and 1,181. Let's check if the length of dictionary as what we expect:

In [22]:
print('Expected length:', len(googlePlay_data) - 1181)
print('Actual length:', len(reviews_max))

Expected length: 9659
Actual length: 9659


Now, let's use **reviews_max** dictionary to remove the duplicates by do the following: 

1- We need to creat two empty list:
 * **googlePlay_clean** list: to store the correct apps(without the duplicates apps).
 * **already_added** list: to store The apps names of the correct apps.

2- We loop through the **googlePlay_data**, and for every iteration, we add the current row(app) to the  **googlePlay_clean** list, and the app name(name) to the **already_added** list if the two following conditions are true:

* Conditions(1):
The **n_reviews** of the current app is equal to the number of reviews of that app in the **reviews_max** dictionary.

* Conditions(2):
The name of the current app is not in the **'already_added'** list , we need this condtion for those cases where the highest number of reviews of a duplicate app is the same for more than one entry and If we only check for the first condition, we'll still end up with duplicate entries for some apps.

In [25]:
googlePlay_clean = []
already_added = []

for app in googlePlay_data:
    name = app[0]
    n_reviews = float(app[3])
    
    if (n_reviews == reviews_max[name]) and (name not in already_added):
        googlePlay_clean.append(app)
        already_added.append(name)


Now let's quickly explore the new data set, and confirm that the number of rows is 9,659:

In [26]:
explore_data(googlePlay_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


We have 9659 rows, just as we expected.

### 2.3 Removing Non-English Apps

If we explore the data enough, we'll find that both data sets have apps with names that suggest they are not directed toward an English-speaking audience. See the below examples:

In [40]:
print(appStore_data[6731][1])
print(googlePlay_clean[4412][0])


【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜
中国語 AQリスニング


We're not interested in keeping these apps, so we'll remove them. One way to go about this is to remove each app with a name containing a symbol that is not commonly used in English text.  English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;), and other symbols (+, *, /).


We know that each character we use in a string has a corresponding number associated with it. The numbers corresponding to the characters we commonly use in an English text are all in the range 0 to 127, according to the ASCII system. Based on this number range, we can build a function that detects whether a character belongs to the set of common English characters or not. If the number is equal to or less than 127, then the character belongs to the set of common English characters.


So, We built this function below using the built-in ord() function to find out the corresponding encoding number of each character.


In [41]:
def is_English(string):
    for character in string:
        if ord(character) > 127:
            return False 
    return True
        

In [42]:
#Check:
print(is_English('Instagram'))
print(is_English('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_English('Docs To Go™ Free Office Suite'))
print(is_English('Instachat 😜'))

True
False
False
False


The function seems to work fine, but some English app names use emojis or other symbols that fall outside of the ASCII range. Because of this, we'll remove useful apps if we use the function in its current form. So,we need to use a function that check if the character is emoji or not.

**Note:**
you'll need to Install [**emoji**](https://github.com/carpedm20/emoji/) package to use [char_is_emoji](https://gist.github.com/jezdez/0185e35704dbdf3c880f) Function. 

In [97]:
import emoji

def char_is_emoji(character):
    return character in emoji.UNICODE_EMOJI

In [98]:
def is_English(string):

    for character in string:
        if ord(character) > 127:
            if (char_is_emoji(character)) != True:
                        return False 
    
    return True

In [100]:
#check again:
print(is_English('Docs To Go™ Free Office Suite'))
print(is_English('Instachat 😜'))

True
True


The function works perfectly ,Now we can use it to filter out the non-English apps for both data sets:

In [101]:
appStore_english = []
googlePlay_english = []

for app in appStore_data:
    name = app[1]
    if is_English(name):
        appStore_english.append(app)

for app in googlePlay_clean:
    name = app[0]
    if is_English(name):
        googlePlay_english.append(app)        
        

In [102]:
explore_data(appStore_english, 0, 3, True)
print('\n')
explore_data(googlePlay_english, 0, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 5874
Number of columns: 16


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '

After removing the non-english apps, we're left with 5874 iOS apps and 9283 Android apps.

### 2.4 Removing Non-Free Apps

As we mentioned, we only need the apps that are free to download and install. So, we'll need to remove non-free apps as the last step in the data cleaning process

In [103]:
appStore_freeApp = []
googlePlay_freeApp = []

for app in appStore_english:
    price = app[4] 
    if price == '0.0' :
        appStore_freeApp.append(app)

for app in googlePlay_english:
    price = app[7]
    if  price == '0':
        googlePlay_freeApp.append(app)  

In [104]:
print('the number of free app in App Store: ' , len(appStore_freeApp))
print('the number of free app in Google Play: ' , len(googlePlay_freeApp))

the number of free app in App Store:  3020
the number of free app in Google Play:  8555


We have 3020 iOS apps and 8555 Android apps, which should be enough for our analysis.

## Data Analysis


### Most Common Apps by Genre

Now, we're going to start analyzing the data:

As we mentioned in the introduction, the aim of this project is to determine the kinds of apps that are likely to attract more users because the revenue is highly influenced by the number of people using the apps.

To minimize risks and overhead, the company uses the validation strategy for an app idea which comprised of 3 steps:

1- Build a minimal Android version of the app, and add it to Google Play.

2- If the app has a good response from users, we develop it further.

3- If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because the company wants to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets. Let's begin the analysis by getting a sense of what are the most common genres for each market.For this, we'll need to build frequency tables for 'prime_genre' column in app Store data set and for 'Genres' and 'Category' columns in google Play data set.We'll build two functions to analyze the frequency tables:

1- One function to generate frequency tables that show percentages.

2- Another function to display the percentages in a descending order.

In [105]:

def freq_table(dataset ,index):
    
    frequency_table = {}
    total = 0

    for row in dataset:
        total += 1
        value = row[index]
        
        if value in frequency_table:
            frequency_table[value]+=1
        else:
            frequency_table[value]= 1
            
    percentages_table = {}
    for key in frequency_table:
        percentage = (frequency_table[key] / total) * 100
        percentages_table[key] = percentage
    
    return percentages_table


        
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])


Let's start by exploring the frequency tables:

1- The frequency table for the 'prime_genre' column of the 'App Store' data set.

In [106]:
display_table(appStore_freeApp, -5 ) #prime_genre

Games : 59.60264900662252
Entertainment : 7.6158940397351
Photo & Video : 4.966887417218543
Education : 3.80794701986755
Social Networking : 3.1788079470198674
Shopping : 2.4503311258278146
Utilities : 2.218543046357616
Music : 2.0860927152317883
Sports : 2.019867549668874
Health & Fitness : 1.95364238410596
Productivity : 1.6556291390728477
Lifestyle : 1.490066225165563
News : 1.2913907284768211
Travel : 1.0927152317880795
Finance : 1.0927152317880795
Weather : 0.8609271523178808
Food & Drink : 0.8609271523178808
Reference : 0.5298013245033113
Business : 0.49668874172185434
Book : 0.26490066225165565
Medical : 0.1986754966887417
Navigation : 0.16556291390728478
Catalogs : 0.09933774834437085


From result, we can see that among the ios apps, more than a half (59.60%) are games. Entertainment apps are close to  8%, followed by photo and video apps, which are close to 5%.  Only 3.8% of the apps are designed for education, followed by social networking apps which amount for 3.18% of the apps in our data set.

The general impression is that App Store (the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking,music, sports, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.


2- The frequency table for the 'Category' column of the 'Google Play' data set.

In [119]:
display_table(googlePlay_freeApp, 1)#Category

FAMILY : 18.92460549386324
GAME : 9.72530683810637
TOOLS : 8.556399766218584
BUSINESS : 4.687317358270017
PRODUCTIVITY : 3.9392168322618355
LIFESTYLE : 3.9041496201052017
FINANCE : 3.717124488603156
MEDICAL : 3.6236119228521333
PERSONALIZATION : 3.378141437755698
SPORTS : 3.237872589129164
COMMUNICATION : 3.2028053769725306
HEALTH_AND_FITNESS : 3.132670952659264
PHOTOGRAPHY : 2.9690239625949735
NEWS_AND_MAGAZINES : 2.77030976037405
SOCIAL : 2.6651081239041496
TRAVEL_AND_LOCAL : 2.2910578609000583
SHOPPING : 2.2209234365867916
BOOKS_AND_REFERENCE : 2.162478082992402
DATING : 1.8234950321449443
VIDEO_PLAYERS : 1.7533606078316772
MAPS_AND_NAVIGATION : 1.3325540619520748
FOOD_AND_DRINK : 1.2273524254821742
EDUCATION : 1.157218001168907
ENTERTAINMENT : 0.9468147282291058
LIBRARIES_AND_DEMO : 0.9351256575102279
AUTO_AND_VEHICLES : 0.9234365867913501
HOUSE_AND_HOME : 0.8182349503214494
WEATHER : 0.783167738164816
EVENTS : 0.701344243132671
PARENTING : 0.6545879602571595
ART_AND_DESIGN : 0.654

The result seems a little different on Google Play: 
- There are not that many apps designed for fun, However, if we investigate this further, we can see that the family category (19% of the apps) means mostly games for kids.

- Also, it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.).  so, practical apps seem to have a better representation on Google Play compared to App Store. This picture is also confirmed by the frequency table we see below for the Genres column:

3- The frequency table for the 'Genres' column of the 'Google Play' data set.

In [108]:
display_table(googlePlay_freeApp, 9)

Tools : 8.544710695499708
Entertainment : 6.054938632378725
Education : 5.341905318527177
Business : 4.687317358270017
Productivity : 3.9392168322618355
Lifestyle : 3.892460549386324
Finance : 3.717124488603156
Medical : 3.6236119228521333
Personalization : 3.378141437755698
Sports : 3.3080070134424315
Communication : 3.2028053769725306
Action : 3.1560490940970194
Health & Fitness : 3.132670952659264
Photography : 2.9690239625949735
News & Magazines : 2.77030976037405
Social : 2.6651081239041496
Travel & Local : 2.2910578609000583
Shopping : 2.2209234365867916
Books & Reference : 2.162478082992402
Simulation : 2.104032729398013
Arcade : 1.858562244301578
Dating : 1.8234950321449443
Casual : 1.8001168907071887
Video Players & Editors : 1.7299824663939218
Maps & Navigation : 1.3325540619520748
Food & Drink : 1.2273524254821742
Puzzle : 1.1455289304500291
Racing : 1.0169491525423728
Strategy : 0.9468147282291058
Role Playing : 0.9468147282291058
Libraries & Demo : 0.9351256575102279
Auto 

The difference between the Genres and the Category columns on Google Play dataset is not clear, but one thing we can notice is that the Genres column seems more detailed (it has more categories) than  the Category column. We're only looking for the bigger picture, so we'll only work with the Category column moving forward.

Until now, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. 

Next, we want to explore kind of apps that has more users to see what kind of app has high demand.

## Most Popular Apps by Genre on the App Store


One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the 'Installs' column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the 'rating_count_tot' column.

Let's start with calculating the average number of user ratings per app genre on the App Store. To do that, we'll need to:

1- Isolateing the apps of each genre using the freq_table() function that we wrote in the previous step.

2- For each genre:
 * Sum up the user ratings for the apps of that genre.
 * Calculating the average number of user ratings by dividing the sum by the number of apps belonging to that genre.

In [123]:
appStore_genres = freq_table(appStore_freeApp , -5)
 
for genre in appStore_genres :
    total = 0 
    len_genre = 0
    
    for app in appStore_freeApp :
        genre_app = app[-5]
        if genre_app == genre:
            total += float(app[5]) # rating_count_tot
            len_genre += 1   
    avg_numOfRatings = total /len_genre
    print(genre, ':', avg_numOfRatings)

Social Networking : 75253.84375
Photo & Video : 29249.766666666666
Games : 22820.230555555554
Music : 55396.01587301587
Reference : 84258.25
Health & Fitness : 19230.86440677966
Weather : 48275.57692307692
Travel : 34115.57575757576
Shopping : 28517.72972972973
News : 23382.17948717949
Navigation : 102592.0
Lifestyle : 17848.51111111111
Entertainment : 14481.995652173913
Food & Drink : 33333.92307692308
Sports : 25382.114754098362
Finance : 26729.090909090908
Education : 6099.417391304348
Productivity : 22842.22
Utilities : 11413.179104477613
Book : 16671.0
Business : 6839.6
Catalogs : 5195.0
Medical : 612.0


From result, the'Navigation' apps is the most popular apps with 102,592 average number of user ratings followed by 'Reference' apps with 84,258, 'Social Networking' apps with 75,253 ,'Music' apps with 55,396 and 'Weather' apps with 48,275 .

Now, let's take a look on the five most popular genres:

In [125]:
for app in appStore_freeApp:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


We can see that the navigation apps is heavily influenced by 'Waze' app and 'Google Maps' app, they both represent a half million of the user reviews.

The same pattern applies to:

1- Social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. 

2- Music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.
 

In [130]:
for app in appStore_freeApp:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
Jishokun-Japanese English Dictionary & Translator : 0


Reference apps have 84,258 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating.

For Weather apps, people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low.

Now, let's move to the Google Play apps.

## Most Popular Apps by Genre on Google Play

For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [131]:
display_table(googlePlay_freeApp, 5) #installs

1,000,000+ : 15.663354763296317
100,000+ : 11.525423728813559
10,000,000+ : 10.461718293395675
10,000+ : 10.391583869082407
1,000+ : 8.416130917592051
100+ : 7.001753360607832
5,000,000+ : 6.756282875511396
500,000+ : 5.5406195207481
50,000+ : 4.769140853302162
5,000+ : 4.500292226767972
10+ : 3.483343074225599
500+ : 3.226183518410286
50,000,000+ : 2.2910578609000583
100,000,000+ : 2.162478082992402
50+ : 1.9403857393337232
5+ : 0.8065458796025715
1+ : 0.5026300409117476
500,000,000+ : 0.2805376972530684
1,000,000,000+ : 0.22209234365867916
0+ : 0.0467562828755114
0 : 0.01168907071887785


One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users. 

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

**Note:**
we'll need to remove the commas and the plus characters from 'install' values and convert it from string to float.


In [133]:
googlePlay_categories = freq_table(googlePlay_freeApp , 1)

for category in googlePlay_categories :
    total = 0 
    len_category = 0
    for app in googlePlay_freeApp :
        category_app = app[1]
        if category_app == category:
            num_Of_Installs = app[5]
            num_Of_Installs = num_Of_Installs.replace('+' , '')
            num_Of_Installs = num_Of_Installs.replace(',' , '')
            num_Of_Installs = float(num_Of_Installs)
            total += num_Of_Installs
            len_category += 1   
    avg_numOfInstalls = total /len_category
    print(category , ':' , avg_numOfInstalls )

 

ART_AND_DESIGN : 1932519.642857143
AUTO_AND_VEHICLES : 645317.2278481013
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8999314.918918919
BUSINESS : 1710203.4663341646
COMICS : 880440.625
COMMUNICATION : 35933961.6459854
DATING : 822459.9807692308
EDUCATION : 1826767.6767676768
ENTERTAINMENT : 12177283.950617284
EVENTS : 232885.83333333334
FINANCE : 1380165.5094339622
FOOD_AND_DRINK : 1911606.2
HEALTH_AND_FITNESS : 4258762.6567164175
HOUSE_AND_HOME : 1380033.7285714287
LIBRARIES_AND_DEMO : 662421.375
LIFESTYLE : 1381354.248502994
GAME : 15799823.725961538
FAMILY : 3696338.7331686225
MEDICAL : 121230.14193548387
SOCIAL : 24038639.263157893
SHOPPING : 7274624.131578947
PHOTOGRAPHY : 18028223.68110236
SPORTS : 3860842.6606498193
TRAVEL_AND_LOCAL : 14370321.867346939
TOOLS : 11010869.950819673
PERSONALIZATION : 5254790.269896193
PRODUCTIVITY : 16852579.56676558
PARENTING : 535196.6071428572
WEATHER : 5219216.7164179105
VIDEO_PLAYERS : 24964878.133333333
NEWS_AND_MAGAZINES : 9926131.2658

From result, 'Communication' apps have the most installs with 35,933,961 followed by 'Video players' apps with 24,964,878 ,'Social' apps with 24,038,639 ,'Photography' apps with 18,028,223 ,'Productivity' apps with 16,852,579  and Game apps with 15,799,823.


Even though that the average number of installs for 'communication apps' is high , but this result is heavily skewed up by a few apps that have over one billion installs(like: WhatsApp Messenger , Skype, Google Chrome, Gmail, Hangouts) ,and a few others with over 100 and 500 million installs:

In [134]:
for app in googlePlay_freeApp:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Messenger : 500,000,000+
WeChat : 100,000,000+
BBM - Free Call

If we removed all the communication apps that have over 100 million installs, the average would be reduced roughly ten times( from 35 million to 3 million):

In [140]:
under_100M = []

for app in googlePlay_freeApp:
    num_Of_Installs = app[5]
    num_Of_Installs = num_Of_Installs.replace(',', '')
    num_Of_Installs = num_Of_Installs.replace('+', '')
    num_Of_Installs = float(num_Of_Installs) 

    if (app[1] == 'COMMUNICATION') and (num_Of_Installs < 100000000):
        under_100M.append(num_Of_Installs)
        
sum(under_100M) / len(under_100M)

3397210.807228916

We see the same pattern repeated for:

1- Video players apps with Youtube, 'Google Play Movies & TV' and MX Player.

2- Socia apps with Facebook, Instagram, Google , etc.

3- Photography apps with Google Photos and other popular photo editors.

4- Productivity apps with Microsoft Word, Dropbox, Google Calendar, Evernote, etc.

This pattern can give us some hints that these app genres may not really popular than they are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

For 'Game' apps, it seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

Our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play. Since we found the 'reference' genre has some potential to work well on the App Store, also on Google Play (BOOKS_AND_REFERENCE) looks fairly popular as well, with an average number of installs of 8,999,314. It's interesting to explore this in more depth, 


In [141]:
for app in googlePlay_freeApp:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])


E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
English translation from Bengali : 100,000

The'BOOKS_AND_REFERENCE' genre includes a variety of apps also It seems there's still a small number of extremely popular apps that skew the average:

In [142]:
for app in googlePlay_freeApp:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])


Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


Luckily, it looks like there are only a few very popular apps(that skew the average), so this market still shows potential. Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [144]:
for app in googlePlay_freeApp:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
Hafizi Quran 15 lines per page : 1,000,000

The 'BOOKS_AND_REFERENCE' genre seems to be dominated by:

1- Softwares for reading ebooks. 

2- Various collections of libraries and dictionaries.

So it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable.

## Conclusions

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book.This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, and many interesting features that will make the user spend more time using the apps and watching more ads which lead to increase the revenue.